File: packages/go/creational.go
// Package gof implements the 22 Gang of Four design patterns as small, typed
// helpers: you supply the domain logic, the helper owns the bookkeeping.
//
// Go has no inheritance, so several patterns are a func or a channel rather than a
// class hierarchy — where the language already does the job, the doc comment says
// so and you should use the language. See the README for the full table.
//
// Nothing here imports anything outside the standard library.
package gof

import (
	"encoding/json"
	"fmt"
	"sync"
)

// Singleton holds one lazily built value, shared.
//
// [sync.OnceValue] is the idiomatic way to do this and you should prefer it. This
// exists for [Singleton.Reset], which a test needs and OnceValue does not offer.
type Singleton[T any] struct {
	factory func() T
	mu      sync.Mutex
	value   T
	built   bool
}

// NewSingleton returns a Singleton that calls factory at most once.
func NewSingleton[T any](factory func() T) *Singleton[T] {
	return &Singleton[T]{factory: factory}
}

// Value builds the value on first call and returns the same one after that. It is
// safe to call from several goroutines.
func (s *Singleton[T]) Value() T {
	s.mu.Lock()
	defer s.mu.Unlock()
	if !s.built {
		s.value = s.factory()
		s.built = true
	}
	return s.value
}

// Reset drops the value, so the next Value call builds a new one.
func (s *Singleton[T]) Reset() {
	s.mu.Lock()
	defer s.mu.Unlock()
	var zero T
	s.value = zero
	s.built = false
}

// Registry holds factories keyed by name: Factory Method, Abstract Factory and
// Strategy are the same code, differing only in what you register.
//
// F is the func type you choose, so the call stays typed:
//
//	rails := gof.NewRegistry[string, func(int) string]()
//	rails.Register("stripe", func(cents int) string { return fmt.Sprint("stripe:", cents) })
//	send, _ := rails.Get("stripe")
//	send(500)
type Registry[K comparable, F any] struct {
	mu        sync.RWMutex
	factories map[K]F
}

// NewRegistry returns an empty Registry.
func NewRegistry[K comparable, F any]() *Registry[K, F] {
	return &Registry[K, F]{factories: make(map[K]F)}
}

// Register stores factory under key and returns it, so a registration can be kept
// in a variable.
func (r *Registry[K, F]) Register(key K, factory F) F {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.factories[key] = factory
	return factory
}

// Get returns the factory for key, or an error naming the key.
func (r *Registry[K, F]) Get(key K) (F, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	factory, ok := r.factories[key]
	if !ok {
		var zero F
		return zero, fmt.Errorf("gof: nothing registered for %v", key)
	}
	return factory, nil
}

// MustGet is Get for the case where a missing key is a programming error.
func (r *Registry[K, F]) MustGet(key K) F {
	factory, err := r.Get(key)
	if err != nil {
		panic(err)
	}
	return factory
}

// Has reports whether key is registered.
func (r *Registry[K, F]) Has(key K) bool {
	r.mu.RLock()
	defer r.mu.RUnlock()
	_, ok := r.factories[key]
	return ok
}

// Keys returns the registered keys, in no particular order.
func (r *Registry[K, F]) Keys() []K {
	r.mu.RLock()
	defer r.mu.RUnlock()
	keys := make([]K, 0, len(r.factories))
	for key := range r.factories {
		keys = append(keys, key)
	}
	return keys
}

// Clone is Prototype: a deep copy, so a new value starts from an existing one.
//
// Go has no generic deep copy, so this round-trips through JSON. That means the
// type has to be JSON-serialisable, and anything JSON cannot carry is dropped:
// funcs, channels, unexported fields, cycles. For a type you own, a Clone method
// is faster and exact — write that.
func Clone[T any](value T) (T, error) {
	var copied T
	encoded, err := json.Marshal(value)
	if err != nil {
		return copied, fmt.Errorf("gof: Clone could not encode: %w", err)
	}
	if err := json.Unmarshal(encoded, &copied); err != nil {
		return copied, fmt.Errorf("gof: Clone could not decode: %w", err)
	}
	return copied, nil
}

// Builder builds a value step by step.
//
// Functional options — New(opts ...Option) — are the idiomatic Go builder and you
// should reach for those first. [Builder.WithIf] is what this adds: construction
// spread across branches, such as a query assembled from optional filters.
type Builder[T any] struct {
	draft T
}

// NewBuilder returns a Builder starting from draft.
func NewBuilder[T any](draft T) *Builder[T] {
	return &Builder[T]{draft: draft}
}

// With applies change to the draft.
func (b *Builder[T]) With(change func(T) T) *Builder[T] {
	b.draft = change(b.draft)
	return b
}

// WithIf applies change only when condition holds.
func (b *Builder[T]) WithIf(condition bool, change func(T) T) *Builder[T] {
	if condition {
		return b.With(change)
	}
	return b
}

// Build returns the value.
func (b *Builder[T]) Build() T { return b.draft }

File: packages/rust/src/creational.rs
//! Creational patterns.
//!
//! Prototype ships no helper: [`Clone`] *is* the pattern, and `#[derive(Clone)]`
//! is shorter than anything wrapping it could be.

use std::collections::HashMap;
use std::fmt;
use std::hash::Hash;
use std::sync::{Arc, RwLock};

/// Singleton — one lazily built value, shared.
///
/// [`std::sync::LazyLock`] is the idiomatic way to do this and you should prefer
/// it. This exists for [`Singleton::reset`], which a test needs and `LazyLock`
/// does not offer.
///
/// Access is by `Arc<T>` rather than `&T`: the value lives behind a lock, so it
/// cannot be lent out as a plain reference.
///
/// ```
/// use gof_patterns::Singleton;
///
/// let calls = std::sync::atomic::AtomicUsize::new(0);
/// let config = Singleton::new(|| {
///     calls.fetch_add(1, std::sync::atomic::Ordering::SeqCst);
///     "MXN".to_string()
/// });
///
/// assert!(Arc::ptr_eq(&config.get(), &config.get()));
/// assert_eq!(calls.load(std::sync::atomic::Ordering::SeqCst), 1);
/// # use std::sync::Arc;
/// ```
pub struct Singleton<T, F = fn() -> T>
where
    F: Fn() -> T,
{
    factory: F,
    value: RwLock<Option<Arc<T>>>,
}

impl<T, F: Fn() -> T> Singleton<T, F> {
    /// Wraps a factory that will be called at most once per build.
    pub fn new(factory: F) -> Self {
        Self {
            factory,
            value: RwLock::new(None),
        }
    }

    /// Builds the value on first call and returns the same `Arc` afterwards. Safe
    /// to call from several threads; the factory runs once even under contention.
    pub fn get(&self) -> Arc<T> {
        if let Some(existing) = self.value.read().expect("Singleton lock poisoned").as_ref() {
            return Arc::clone(existing);
        }
        let mut slot = self.value.write().expect("Singleton lock poisoned");
        // Another thread may have built it between the read and the write.
        if let Some(existing) = slot.as_ref() {
            return Arc::clone(existing);
        }
        let created = Arc::new((self.factory)());
        *slot = Some(Arc::clone(&created));
        created
    }

    /// Drops the value, so the next [`Singleton::get`] builds a new one.
    pub fn reset(&self) {
        *self.value.write().expect("Singleton lock poisoned") = None;
    }
}

/// The error [`Registry::get`] returns, naming the key it did not know.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RegistryError {
    /// The key, formatted with [`std::fmt::Debug`], because `K` need not be `Display`.
    pub key: String,
}

impl fmt::Display for RegistryError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "gof: nothing registered for {}", self.key)
    }
}

impl std::error::Error for RegistryError {}

/// Factory Method, Abstract Factory and Strategy — factories keyed by name. The
/// same code; only what you register differs.
///
/// `F` is the type you choose, usually a boxed closure, so the call stays typed:
///
/// ```
/// use gof_patterns::Registry;
///
/// let mut rails: Registry<&str, Box<dyn Fn(u32) -> String>> = Registry::new();
/// rails.register("stripe", Box::new(|cents| format!("stripe:{cents}")));
///
/// assert_eq!(rails.get("stripe").unwrap()(500), "stripe:500");
/// assert!(rails.get("paypal").is_err());
/// ```
pub struct Registry<K: Eq + Hash + fmt::Debug, F> {
    factories: HashMap<K, F>,
}

impl<K: Eq + Hash + fmt::Debug, F> Registry<K, F> {
    /// An empty registry.
    pub fn new() -> Self {
        Self {
            factories: HashMap::new(),
        }
    }

    /// Stores `factory` under `key`, replacing anything already there.
    pub fn register(&mut self, key: K, factory: F) {
        self.factories.insert(key, factory);
    }

    /// The factory for `key`, or a [`RegistryError`] naming it.
    pub fn get(&self, key: K) -> Result<&F, RegistryError> {
        self.factories.get(&key).ok_or_else(|| RegistryError {
            key: format!("{key:?}"),
        })
    }

    /// Whether `key` is registered.
    pub fn has(&self, key: &K) -> bool {
        self.factories.contains_key(key)
    }

    /// The registered keys, in no particular order.
    pub fn keys(&self) -> impl Iterator<Item = &K> {
        self.factories.keys()
    }

    /// How many factories are registered.
    pub fn len(&self) -> usize {
        self.factories.len()
    }

    /// Whether nothing is registered.
    pub fn is_empty(&self) -> bool {
        self.factories.is_empty()
    }
}

impl<K: Eq + Hash + fmt::Debug, F> Default for Registry<K, F> {
    fn default() -> Self {
        Self::new()
    }
}

/// Builder — construction spread across branches.
///
/// A hand-written builder with typed setters is more idiomatic when the fields are
/// known; [`Builder::with_if`] is what this adds, for a query or a request
/// assembled from optional parts. It takes `self` by value, so the chain moves
/// rather than borrowing.
///
/// ```
/// use gof_patterns::Builder;
///
/// #[derive(Debug, PartialEq, Default, Clone)]
/// struct Query { table: String, user: Option<String> }
///
/// let user = Some("u1".to_string());
/// let query = Builder::new(Query::default())
///     .with(|mut q| { q.table = "orders".into(); q })
///     .with_if(user.is_some(), |mut q| { q.user = user.clone(); q })
///     .build();
///
/// assert_eq!(query.table, "orders");
/// assert_eq!(query.user.as_deref(), Some("u1"));
/// ```
pub struct Builder<T> {
    draft: T,
}

impl<T> Builder<T> {
    /// Starts from `draft`.
    pub fn new(draft: T) -> Self {
        Self { draft }
    }

    /// Applies `change`.
    #[must_use]
    pub fn with(mut self, change: impl FnOnce(T) -> T) -> Self {
        self.draft = change(self.draft);
        self
    }

    /// Applies `change` only when `condition` holds.
    #[must_use]
    pub fn with_if(self, condition: bool, change: impl FnOnce(T) -> T) -> Self {
        if condition {
            self.with(change)
        } else {
            self
        }
    }

    /// Hands back the value.
    pub fn build(self) -> T {
        self.draft
    }
}
